applications: nrf_desktop: Add buttons documentation (#1133)
NordicPlayground/fw-nrfconnect-nrf: applications/nrf_desktop/doc/buttons.rst

``buttons`` module is responsible for generating events related to key presses.
The source of events are changes to GPIO pins.

Module Events
*************

+----------------+-------------+----------------+------------------+-------------------------+
| Source Module  | Input Event | This Module    | Output Event     | Sink Module             |
+================+=============+================+==================+=========================+
|                |             | :ref:`buttons` | ``button_event`` | :ref:`hid_state`        |
+                +             +                +                  +-------------------------+
|                |             |                |                  | :ref:`fn_keys`          |
+                +             +                +                  +-------------------------+
|                |             |                |                  | :ref:`click_detector`   |
+----------------+-------------+----------------+------------------+-------------------------+

Configuration
*************

Module is enabled by ``CONFIG_DESKTOP_BUTTONS_ENABLE``.

Module can handle both matrix keyboard and buttons connected directly to GPIO pins.

> When defining how button are connected two arrays are used. First one contains

Author: fixed

pins associated with matrix rows, second pins associated with columns. Both
arrays are defined in file ``buttons_def.h`` that is located in the board
specific directory in application configuration folder. Array of column pins
can be left empty in which cases buttons are assumed to be directly connected
to row pins (one button per pin).

By default it is assumed that button press is indicated by pin switch from low
to high state. This can be changed with ``CONFIG_DESKTOP_BUTTONS_POLARITY_INVERSED``
that will cause application react to pin change from high to low state.

Implementation details
**********************

``buttons`` can be in the following states:

> 

Author: fixed

>     * STATE_IDLE

Author: done

    * STATE_ACTIVE
    * STATE_SCANNING
    * STATE_SUSPENDING

In ``STATE_ACTIVE`` module enables GPIO interrupts and wait for pins state to
change. When any button is pressed module switches to ``STATE_SCANNING``.
When switch occurs module submits a work with a delay set to
``CONFIG_DESKTOP_BUTTONS_DEBOUNCE_INTERVAL``. Work performs scanning of the
keyboard matrix noting changes to button states and sending related events.

> While any button is kept pressed while scanning is performed work will be

Author: Yes, but scanning will continue while any button is pressed. E.g. we press A -> scanning is started and performed every T time. If A is released but B is pressed between scans, scanning will continue. If there is no button pressed while scanning, scan will stop and module will switch to an interrupt based mode.

re-submitted with delay set to ``CONFIG_DESKTOP_BUTTONS_SCAN_INTERVAL``.
If no button is pressed module switch back to ``STATE_ACTIVE``.

> When system enters low-power state, ``buttons`` goes to ``STATE_IDLE`` in which

Reviewer: ```suggestion
When the system enters the low-power state, the ``buttons`` module goes to ``STATE_IDLE``, in which
```

> it waits for GPIO interrupts indicating change to button states. When interrupt

Reviewer: ```suggestion
it waits for GPIO interrupts that indicate a change to button states. When an interrupt
```

> is triggered module will issue a system wake up event.

Reviewer: ```suggestion
is triggered, the module will issue a system wakeup event.
```

> In case request for power off comes while button is held, module switches to

Reviewer: ```suggestion

If a request for power off comes while the button is pressed, the module switches to
```
Move this paragraph to a new paragraph after line 59.

> ``STATE_SUSPENDING`` and be in this state until no button is pressed. Then it

Reviewer: ```suggestion
``STATE_SUSPENDING`` and remains in this state until no button is pressed. Then, it
```
